File: 2023/day17/cpp/solution.cpp
/**
 * Day 17: Clumsy Crucible
 * Dijkstra's shortest path with movement constraints.
 *
 * Optimized: uses static array for visited states instead of hash set.
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <queue>
#include <cstring>
#include <string>

using namespace std;

constexpr int MAX_SIZE = 150;
constexpr int MAX_CONSEC = 11;  // 0-10
constexpr int NUM_DIRS = 4;

// Direction deltas: 0=right, 1=down, 2=left, 3=up
constexpr int DR[] = {0, 1, 0, -1};
constexpr int DC[] = {1, 0, -1, 0};

// Static visited array: [row][col][dir][consec]
static bool visited[MAX_SIZE][MAX_SIZE][NUM_DIRS][MAX_CONSEC];

struct State {
    int heat;
    int row;
    int col;
    int dir;
    int consec;

    bool operator>(const State& other) const {
        return heat > other.heat;
    }
};

vector<vector<int>> parseInput(const string& filename) {
    ifstream file(filename);
    vector<vector<int>> grid;
    string line;

    while (getline(file, line)) {
        if (line.empty()) continue;
        vector<int> row;
        for (char c : line) {
            row.push_back(c - '0');
        }
        grid.push_back(row);
    }

    return grid;
}

int dijkstra(const vector<vector<int>>& grid, int minStraight, int maxStraight) {
    int rows = grid.size();
    int cols = grid[0].size();

    // Clear visited array
    memset(visited, 0, sizeof(visited));

    // Priority queue: min-heap by heat loss
    priority_queue<State, vector<State>, greater<State>> pq;

    // Start with no direction (-1)
    pq.push({0, 0, 0, -1, 0});

    while (!pq.empty()) {
        State curr = pq.top();
        pq.pop();

        // Check if we reached the goal
        if (curr.row == rows - 1 && curr.col == cols - 1) {
            if (minStraight == 0 || curr.consec >= minStraight) {
                return curr.heat;
            }
        }

        // Check if already visited (skip dir=-1 case)
        if (curr.dir >= 0 && visited[curr.row][curr.col][curr.dir][curr.consec]) {
            continue;
        }
        if (curr.dir >= 0) {
            visited[curr.row][curr.col][curr.dir][curr.consec] = true;
        }

        // Try all four directions
        for (int nd = 0; nd < 4; nd++) {
            // Can't reverse direction
            if (curr.dir != -1 && nd == (curr.dir + 2) % 4) {
                continue;
            }

            int nr = curr.row + DR[nd];
            int nc = curr.col + DC[nd];

            // Bounds check
            if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) {
                continue;
            }

            int newConsec;
            if (nd == curr.dir) {
                // Continuing in same direction
                newConsec = curr.consec + 1;
                if (newConsec > maxStraight) {
                    continue;
                }
            } else {
                // Turning - must have gone minStraight in previous direction first
                if (curr.dir != -1 && curr.consec < minStraight) {
                    continue;
                }
                newConsec = 1;
            }

            // Skip if already visited
            if (visited[nr][nc][nd][newConsec]) {
                continue;
            }

            int newHeat = curr.heat + grid[nr][nc];
            pq.push({newHeat, nr, nc, nd, newConsec});
        }
    }

    return -1; // No path found
}

int part1(const vector<vector<int>>& grid) {
    return dijkstra(grid, 0, 3);
}

int part2(const vector<vector<int>>& grid) {
    return dijkstra(grid, 4, 10);
}

int main() {
    auto grid = parseInput("../input.txt");

    cout << "Part 1: " << part1(grid) << endl;
    cout << "Part 2: " << part2(grid) << endl;

    return 0;
}

File: 2023/day17/rust/solution.rs
// Day 17: Clumsy Crucible - Dijkstra's shortest path with movement constraints
//
// Optimized: uses fixed-size array for visited states instead of HashSet.

use std::collections::BinaryHeap;
use std::cmp::Ordering;
use std::fs;
use std::path::Path;

const MAX_SIZE: usize = 150;
const MAX_CONSEC: usize = 11;  // 0-10
const NUM_DIRS: usize = 4;

// Direction deltas: right, down, left, up
const DR: [i32; 4] = [0, 1, 0, -1];
const DC: [i32; 4] = [1, 0, -1, 0];

#[derive(Clone, Copy, Eq, PartialEq)]
struct Node {
    heat: u32,
    row: i32,
    col: i32,
    direction: i8,    // -1 = no direction, 0=right, 1=down, 2=left, 3=up
    consecutive: u8,
}

// For BinaryHeap - we want min-heap, so reverse the ordering
impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        other.heat.cmp(&self.heat)
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn parse_input(filename: &Path) -> Vec<Vec<u8>> {
    let content = fs::read_to_string(filename).expect("Failed to read input file");
    content
        .lines()
        .filter(|line| !line.is_empty())
        .map(|line| line.chars().map(|c| c.to_digit(10).unwrap() as u8).collect())
        .collect()
}

fn dijkstra(grid: &[Vec<u8>], min_straight: u8, max_straight: u8) -> u32 {
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;

    // Visited array: [row][col][dir][consec]
    let mut visited = vec![[[false; MAX_CONSEC]; NUM_DIRS]; MAX_SIZE * MAX_SIZE];

    let mut pq = BinaryHeap::new();

    // Start with no direction
    pq.push(Node {
        heat: 0,
        row: 0,
        col: 0,
        direction: -1,
        consecutive: 0,
    });

    while let Some(Node { heat, row, col, direction, consecutive }) = pq.pop() {
        // Check if we reached the goal
        if row == rows - 1 && col == cols - 1 {
            if min_straight == 0 || consecutive >= min_straight {
                return heat;
            }
        }

        // Check if already visited (skip dir=-1 case)
        if direction >= 0 {
            let idx = (row as usize) * MAX_SIZE + (col as usize);
            if visited[idx][direction as usize][consecutive as usize] {
                continue;
            }
            visited[idx][direction as usize][consecutive as usize] = true;
        }

        // Try all four directions
        for nd in 0..4i8 {
            // Can't reverse direction
            if direction != -1 && nd == (direction + 2) % 4 {
                continue;
            }

            let nr = row + DR[nd as usize];
            let nc = col + DC[nd as usize];

            // Bounds check
            if nr < 0 || nr >= rows || nc < 0 || nc >= cols {
                continue;
            }

            // Check consecutive constraints
            let new_consec: u8;
            if nd == direction {
                // Continuing in same direction
                new_consec = consecutive + 1;
                if new_consec > max_straight {
                    continue;
                }
            } else {
                // Turning - must have gone min_straight in previous direction first
                if direction != -1 && consecutive < min_straight {
                    continue;
                }
                new_consec = 1;
            }

            // Skip if already visited
            let nr_idx = (nr as usize) * MAX_SIZE + (nc as usize);
            if visited[nr_idx][nd as usize][new_consec as usize] {
                continue;
            }

            let new_heat = heat + grid[nr as usize][nc as usize] as u32;
            pq.push(Node {
                heat: new_heat,
                row: nr,
                col: nc,
                direction: nd,
                consecutive: new_consec,
            });
        }
    }

    0 // No path found
}

fn part1(grid: &[Vec<u8>]) -> u32 {
    dijkstra(grid, 0, 3)
}

fn part2(grid: &[Vec<u8>]) -> u32 {
    dijkstra(grid, 4, 10)
}

fn main() {
    let input_path = Path::new(file!()).parent().unwrap().join("../input.txt");
    let grid = parse_input(&input_path);

    println!("Part 1: {}", part1(&grid));
    println!("Part 2: {}", part2(&grid));
}
